OpenRA.Mods.Common/Traits/World/ResourceLayer.cs

```diff
@@ -99,15 +109,45 @@ public void WorldLoaded(World w, WorldRenderer wr)
 					var temp = content[cell];
 					temp.Density = Math.Max(density, 1);
 
-					render[cell] = content[cell] = temp;
-					UpdateRenderedSprite(cell);
+					content[cell] = temp;
+					dirty.Add(cell);
+				}
+			}
+
+			// Build the sprite layer dictionary for rendering resources
+			// All resources that have the same palette must also share a sheet and blend mode
+			foreach (var r in resources)
+			{
+				var pal = r.Value.Palette;
+				TerrainSpriteLayer layer;
+
+				var sheet = r.Value.Variants.First().Value.First().Sheet;
+				var blendMode = r.Value.Variants.First().Value.First().BlendMode;
+
+				if (spriteLayers.TryGetValue(pal, out layer))
+				{
+					sheet = layer.Sheet;
+					blendMode = layer.BlendMode;
+				}
+				else
+				{
+					layer = new TerrainSpriteLayer(w, wr, sheet, blendMode, pal, wr.World.Type != WorldType.Editor);
+					spriteLayers.Add(pal, layer);
 				}
+
+				// Validate that sprites are compatible with this layer
+				if (r.Value.Variants.Any(kv => kv.Value.Any(s => s.Sheet != sheet)))
+					throw new InvalidDataException("Resource sprites span multiple sheets. Try loading their sequences earlier.");
```

Comment:
There are a couple of places where we do checks like this now.  It's out of scope for this PR, but a future one can introduce lint rules to detect these cases.